c7n/filters/cost.py

```diff
@@ -0,0 +1,117 @@
+import os
+
+from c7n.cache import NullCache
+from gql import Client, gql
+from gql.transport.requests import RequestsHTTPTransport
+
+from .core import OPERATORS, Filter
+
+
+class Cost(Filter):
+    """Annotate resource monthly cost with Infracost pricing API.
+    It aims to provide an approximate cost for a generic case. For example,
+    it only grabs the on-demand price with no pre-installed software for EC2 instances.
+
+    Please use INFRACOST_API_ENDPOINT and INFRACOST_API_KEY environment vars
+    to specify the API config.
+
+    .. code-block:: yaml
+
+    policies:
+      - name: ec2-cost
+        resource: ec2
+        filters:
+          - type: cost
+            op: greater-than
+            # USD
+            value: 4
+            # monthly price = unit price * 730 hours
+            quantity: 730
```

Comment:
this is going to vary widely I suspect, (s3, sqs, ebs, versus hour pricing) we'll probably need resource specific examples

Reply:
Right. I am not sure how could this filter end up when comes to more and more resource types. The more I know about the Infracost, the more I hesitate on the filter.

**In order to get the cost information, is the Infracost integration a good solution?**
The resource has been created/billed. It must have some other ways to get the cost info, e.g. CUR, Cost Explorer, etc. I'd like to hear any better alternative solution.
The CUR is too complicated for me. And I don't have an appropriate tag to call the CE api to get cost for an individual resource. Please correct me if I am wrong.

**If Infracost is right, how to integrate with it better?**
The current solution would end up re-implement [quite a lot of the Infracost logic](https://github.com/infracost/infracost/tree/master/internal/resources/aws). I want reuse, not repeat, but I don't know how to achieve that.
Comparing to the Infracost implementation, we will see that the ec2/rds cost filter are missing lots of details.

Reply:
# re is infracost a good solution

I think there is a lot of merit for infracost wrt for static line price on resources, for usage its a bit less clear. I'm game for adding in the integration. cur is complicated but and potentially problematic on query (needs resource information and an Athena setup), agreed that cost explorer isn't going to work as its grouped by billing tags. I've also explored budgets but that's more for account level or resource group costing. 

# re infracost integration

afaics this is the best way, we'll probably need to incorporate metrics at some point as well if we want to try and incorporate usage, for now focusing on resources with structural costs seems fine.

Reply:
Thanks for the comment.